Configure helix-front Maven project to install yarn

micahstubbs opened the pull request:
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

fix #2114 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

This PR configures the helix-front Maven project to use the Javascript package manager [Yarn](https://yarnpkg.org).

This is accomplished by upgrading the frontend-maven-plugin to the [latest version](https://github.com/eirslett/frontend-maven-plugin/tags) and following these instructions in that project's docs: https://github.com/eirslett/frontend-maven-plugin#installing-node-and-yarn

This PR is a hotfix for the build issue observed by @junkaixue on Wednesday 6/8.

### Tests

- [x] The following tests are written for this issue:

(List the names of added unit/integration tests)

- The following is the result of the "mvn test" command on the appropriate module:

(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)

```bash
mvn clean install -Dmaven.test.skip.exec=true

[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Helix 1.0.4-SNAPSHOT:
[INFO]
[INFO] Apache Helix ....................................... SUCCESS [  3.706 s]
[INFO] Apache Helix :: Metrics Common ..................... SUCCESS [  4.567 s]
[INFO] Apache Helix :: Metadata Store Directory Common .... SUCCESS [  2.924 s]
[INFO] Apache Helix :: ZooKeeper API ...................... SUCCESS [  6.383 s]
[INFO] Apache Helix :: Helix Common ....................... SUCCESS [  2.001 s]
[INFO] Apache Helix :: Core ............................... SUCCESS [ 18.490 s]
[INFO] Apache Helix :: Admin Webapp ....................... SUCCESS [  2.987 s]
[INFO] Apache Helix :: Restful Interface .................. SUCCESS [  4.414 s]
[INFO] Apache Helix :: Distributed Lock ................... SUCCESS [  1.506 s]
[INFO] Apache Helix :: HelixAgent ......................... SUCCESS [  1.430 s]
[INFO] Apache Helix :: Front End .......................... SUCCESS [ 30.465 s]
[INFO] Apache Helix :: Recipes ............................ SUCCESS [  0.071 s]
[INFO] Apache Helix :: Recipes :: Rabbitmq Consumer Group . SUCCESS [  2.308 s]
[INFO] Apache Helix :: Recipes :: Rsync Replicated File Store SUCCESS [  1.854 s]
[INFO] Apache Helix :: Recipes :: distributed lock manager  SUCCESS [  1.157 s]
[INFO] Apache Helix :: Recipes :: distributed task execution SUCCESS [  0.901 s]
[INFO] Apache Helix :: Recipes :: service discovery ....... SUCCESS [  1.157 s]
[INFO] Apache Helix :: View Aggregator .................... SUCCESS [  1.524 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:29 min
[INFO] Finished at: 2022-06-09T09:05:45-07:00
[INFO] ------------------------------------------------------------------------
```


(Link the GitHub wiki you added)

### Commits

-[x] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. I

### Code Quality

-[x] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

micahstubbs commented:
I just ran this locally:

```
cd  helix/helix-front
mvn test
```

and see this output:

```py
[INFO] --- maven-compiler-plugin:2.5.1:testCompile (default-testCompile) @ helix-front ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M3:test (default-test) @ helix-front ---
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/3.0.0-M3/surefire-junit3-3.0.0-M3.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/3.0.0-M3/surefire-junit3-3.0.0-M3.jar (21 kB at 24 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/3.0.0-M3/surefire-junit3-3.0.0-M3.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/surefire-junit3/3.0.0-M3/surefire-junit3-3.0.0-M3.pom (1.7 kB at 11 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/common-junit3/3.0.0-M3/common-junit3-3.0.0-M3.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/common-junit3/3.0.0-M3/common-junit3-3.0.0-M3.pom (1.6 kB at 9.7 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/common-junit3/3.0.0-M3/common-junit3-3.0.0-M3.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/surefire/common-junit3/3.0.0-M3/common-junit3-3.0.0-M3.jar (12 kB at 68 kB/s)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  26.980 s
[INFO] Finished at: 2022-06-09T09:12:26-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-front: suiteXmlFiles is configured, but there is no TestNG dependency -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

micahstubbs commented:
:bulb: I think the issue is that the project helix-front has not Maven tests defined at all.  The solution in this case is to run mvn test from the helix root directory, not from helix/helix-front.

micahstubbs commented:
Here's the relevant part of the CI Build `build / build (pull_request)` failure log:

```py
Error:  Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.12.1:yarn (yarn install) on project helix-front: Failed to run task: 'yarn install' failed. org.apache.commons.exec.ExecuteException: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
Error: 
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error: 
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
Error: 
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :helix-front
Error: Process completed with exit code 1.
```

micahstubbs commented:
Ah hah, it looks like the `yarn install` execution is running before the `install node and yarn` execution is complete. 

https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#a-build-phase-is-made-up-of-plugin-goals

> (Note: In Maven 2.0.5 and above, multiple goals bound to a phase are executed in the same order as they are declared in the POM, however multiple instances of the same plugin are not supported. Multiple instances of the same plugin are grouped to execute together and ordered in Maven 2.0.11 and above).

micahstubbs reviewed (commented):
Self-review complete. LGTM.

junkaixue commented:
Did you enable helix-front build in pom.xml of the root? I dont see that change even. Usually helix-front build will take 30 mins not 30s...

micahstubbs commented:
HI @junkaixue , I am confused: didn't you enable `helix-front` in the pom.xml of the root in this commit?

https://github.com/apache/helix/commit/1a91797068a71a4bbcdf5c2c4b06e3b62b3f2945

micahstubbs commented on helix-front/pom.xml:
```diff
@@ -58,7 +58,7 @@
         <groupId>com.github.eirslett</groupId>
         <artifactId>frontend-maven-plugin</artifactId>
-        <version>0.0.22</version>
+        <version>1.12.1</version>
```
use latest version to ensure that we have support for `install-node-and-yarn` goal shown below

micahstubbs replied:
[latest version](https://github.com/eirslett/frontend-maven-plugin/tags) source

micahstubbs commented on helix-front/pom.xml:
```diff
@@ -92,13 +94,13 @@
             </goals>
             <configuration>
-              <arguments>run-script build</arguments>
+              <arguments>build</arguments>
```
The syntax is different between yarn and npm.  https://classic.yarnpkg.com/en/docs/cli/run

Ah! it looks like this should be `run build`.  Will fix.

micahstubbs commented:
This is the error that appears to cause the CI build to fail:

> [INFO] error An unexpected error occurred: "/home/runner/work/helix/helix/helix-front/node_modules/vis-network: Cannot read property 'config' of undefined".
[INFO] info If you think this is a bug, please open a bug report with the information provided in "/home/runner/.config/yarn/global/yarn-error.log".
[INFO] info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

https://github.com/apache/helix/runs/6819388259?check_suite_focus=true#step:4:26219

I am working to try to reproduce the locally.

micahstubbs commented:
I have made two small changes that will hopefully fix the CI build. It turns out we had an old library that we were still importing a stylesheet from. I have removed the old library and updated the style sheet import to refer to the correct new library.
